### ISSUE TITLE
It's possible to register the same event type twice, with extreme instability as a result

### ISSUE THREAD

**Issue by TobiasFella** [quotient-im/libQuotient#413]
One of our users reported this crash. Looks to me like it's not something neochat is doing wrong
Originally reported as https://invent.kde.org/network/neochat/-/issues/4

Built with  libquotient 0.7.0.r25.g12f1eea-1

Just attempted to join any room

```qml: 1
Connected to room "!SEgsRQLScqPxYtucHl:archlinux.org" as "@derfenix:matrix.org"
Both point size and pixel size set. Using pixel size.
file:///usr/lib/qt/qml/org/kde/kirigami.2/private/RefreshableScrollView.qml:137:13: QML Connections: Implicitly defined onFoo properties in Connections are deprecated. Use this syntax instead: function onFoo(<arguments>) { ... }
file:///usr/lib/qt/qml/org/kde/kirigami.2/templates/AbstractApplicationHeader.qml:153:9: QML Connections: Implicitly defined onFoo properties in Connections are deprecated. Use this syntax instead: function onFoo(<arguments>) { ... }
file:///usr/lib/qt/qml/org/kde/kirigami.2/templates/AbstractApplicationHeader.qml:108:9: QML Connections: Implicitly defined onFoo properties in Connections are deprecated. Use this syntax instead: function onFoo(<arguments>) { ... }
file:///usr/lib/qt/qml/org/kde/kirigami.2/templates/AbstractApplicationHeader.qml:83:5: QML Connections: Implicitly defined onFoo properties in Connections are deprecated. Use this syntax instead: function onFoo(<arguments>) { ... }
file:///usr/lib/qt/qml/org/kde/kirigami.2/templates/AbstractApplicationHeader.qml:102:9: QML Connections: Implicitly defined onFoo properties in Connections are deprecated. Use this syntax instead: function onFoo(<arguments>) { ... }
file:///usr/lib/qt/qml/org/kde/kirigami.2/templates/AbstractApplicationHeader.qml:108:9: QML Connections: Implicitly defined onFoo properties in Connections are deprecated. Use this syntax instead: function onFoo(<arguments>) { ... }
file:///usr/lib/qt/qml/org/kde/kirigami.2/templates/AbstractApplicationHeader.qml:102:9: QML Connections: Implicitly defined onFoo properties in Connections are deprecated. Use this syntax instead: function onFoo(<arguments>) { ... }
39 user(s) in the room
ASSERT: "evt" in file /usr/include/Quotient/room.h, line 503
zsh: abort (core dumped)  neochat
```

**Comment by KitsuneRal** [comment]
That's a very unusual location for a failure. Basically, it triggers when the room's cached state is corrupted to the point where a request for an event of some type returns an event of a different type. Which compiler (and version) was this compiled with? Was linking dynamic or static? Also, setting `QT_LOGGING_RULES` to `quotient.*.debug=true;quotient.jobs.sync.debug=false;quotient.events.ephemeral.debug=false` would help with tracking this down (warning, may generate a lot of logs).

**Comment by derfenix** [comment]
Full output with QT_LOGGING_RULES https://gist.github.com/derfenix/75b64a0c1a8aec4136f46c27a8961a5e
Spaces separate logs before room open and after click on room.

The library build commands 
```
cmake ../${pkgname} \
        -DCMAKE_INSTALL_PREFIX=/usr \
        -DCMAKE_INSTALL_LIBDIR=lib \
        -DCMAKE_BUILD_TYPE=Release \
        -DBUILD_SHARED_LIBS=1

make
```

**Comment by derfenix** [comment]
g++ (GCC) 10.2.0
cmake version 3.18.4

**Comment by KitsuneRal** [comment]
Thanks; the logs make the situation quite clear. It looks that I left a loophole for static initialisation fiasco.

The type system for Matrix events in Quotient is extensible: you can define a new event type outside of the library and as long as you have it properly registered (using `DEFINE_EVENT_TYPE_ID` and `REGISTER_EVENT_TYPE` macros) the library will create objects of the registered type when it finds an event of a given type in, say, `/sync` responses. What the logs show is that factories for _most_ event types are registered twice, producing two "valid" class identities for one and the same class and causing extreme instability of the code. The factories are added by means of initialising a static variable with a name bound to the event type. It looks like some of those variables somehow get initialised twice.

With no regard to the reasons of such situation, it's a bug in the current event type registration code. I _think_ I can partially fix it even in 0.6.x; and I certainly can fix it in master that doesn't have ABI/API guarantees. I'd be interested to learn how the same static variable ends up getting initialised twice in your system but that's not a necessity.

**Comment by KitsuneRal** [comment]
It's actually worse because not only factories get added twice but type ids get initialised twice too. The thing with type ids though is that they are made unique for a given type by the virtue of C++ templates. @derfenix, by now I really wonder what's happening on your system that you get a template instantiated twice seemingly for the same C++ type. That is not normal. I can imagine only one case, when header files get included from inside a namespace in one place and outside in the other but then neochat would very likely not even compile.

**Comment by derfenix** [comment]
I did nothing extra, but install libquotient via PKGBUILD from here https://aur.archlinux.org/packages/libquotient-git/ and create own for neochat https://aur.archlinux.org/packages/neochat-git/. 

I'll try to debug this thing if you say how. I'm not familiar with qt or c++, but may be can help somehow.

**Comment by KitsuneRal** [comment]
Let's start with building and running `quotest` target from the libQuotient source tree - it's a small command-line functional test for the library. It will ask for username and password as the first two parameters - provide any workable; the third parameter is the device name (any will do); and the fourth parameter is the alias or id of the test room (you can use `#quotest:matrix.org`). The fifth parameter is the "origin" - just a string that it prepends to its messages in the room. If everything's right, it will join the room you passed and will run some tests on it. This will show clearly whether neochat is relevant to the problem, at least.

**Comment by derfenix** [comment]
Nothing happens after Test room name
```
./quotest derfenix@matrix.org password "#quotest:matrix.org" 123                                                                                                                                                                                                [master:~/tmp/libquotient-git/src/libquotient-git/tests][pts/1]
Connecting to Matrix as derfenix@matrix.org
Test room name: 123

^C
```

**Comment by KitsuneRal** [comment]
You forgot the device name.

**Comment by derfenix** [comment]
Oh, sorry.. 
But not much better result :) 

```
./quotest derfenix@matrix.org password fenixpc "#quotest:matrix.org" test                                                                                                                                                                                       [master:~/tmp/libquotient-git/src/libquotient-git/tests][pts/1]
Connecting to Matrix as derfenix@matrix.org
Test room name: #quotest:matrix.org
Origin for the test message: test
```
And stuck again

**Comment by KitsuneRal** [comment]
Can you increase log level in the same way as above?

**Comment by derfenix** [comment]
Some minutes later (more than 5) exit with 
```
Logging out
quotient.jobs: Developers, ensure the Connection is valid before using it
quotient.jobs: "LogoutJob" status IncorrectRequest: Invalid server connection
quotient.jobs: "LogoutJob" stopped with empty network reply
quotient.jobs: "LogoutJob" stopped with empty network reply
```

> Can you increase log level in the same way as above?

Moment.

**Comment by derfenix** [comment]
My fault, sorry..

**Comment by KitsuneRal** [comment]
Oh - just noticed: user name shouldn't be an email address, it should be something like `@derfenix:matrix.org`.

**Comment by derfenix** [comment]
Bad login string :( 

Tests running now

**Comment by derfenix** [comment]
> Oh - just noticed: user name shouldn't be an email address, it should be something like `@derfenix:matrix.org`.

Yes, I'm not often used matrix before and just miss this thing, sorry again :( 

Here is the tests result if it required https://gist.github.com/derfenix/14631456ad0461d7b0f1a5399593b9da

**Comment by KitsuneRal** [comment]
The output looks quite right. The one failing test is ok and not relevant to this issue. Did you build it with `BUILD_SHARED_LIBRARY=1` or linked the test application statically?

**Comment by derfenix** [comment]
Did run cmake with no additional params, `ldd quotest` shows no libquotient in list

**Comment by KitsuneRal** [comment]
Try building with dynamic linking - just to check whether it is (not) to blame.

**Comment by derfenix** [comment]
Can you specify exact parameters should I add to build with dynamic linking?

**Comment by KitsuneRal** [comment]
If you pass `BUILD_SHARED_LIBS=1` to cmake for libQuotient, you should only be getting the dynamic library in the build directory. It makes sense to do `cmake --target clean` (aka `make clean` if you use `make` for the buildtool) before building the library, to be sure there's no `libQuotient.a` lying around.

**Comment by derfenix** [comment]
```
cmake --target clean .
make clean
cmake -DBUILD_SHARED_LIBS=1 .
make
./quotest @derfenix:matrix.org password fenixpc "#quotest:matrix.org"
```

https://gist.github.com/derfenix/6680f55bfb436d6c850d762805b5f644

**Comment by derfenix** [comment]
```
ls                                                                                                                                                                                                                                                                  
CMakeCache.txt  CMakeFiles/  cmake_install.cmake  CMakeLists.txt  Makefile  quotest*  quotest_autogen/  quotest.cpp
```

**Comment by KitsuneRal** [comment]
Well, this `ls` is not really telling anything because it's the directory of `quotest`. What does ldd say?

**Comment by derfenix** [comment]
```
ldd quotest                                                                                                                                                                                                                                                         [master:~/tmp/libquotient-git/src/libquotient-git/tests][pts/3]
        linux-vdso.so.1 (0x00007ffc851bc000)
        libQt5Test.so.5 => /usr/lib/libQt5Test.so.5 (0x00007f0915d42000)
        libQuotient.so.0.7 => /usr/lib/libQuotient.so.0.7 (0x00007f0915bb9000)
        libQt5Multimedia.so.5 => /usr/lib/libQt5Multimedia.so.5 (0x00007f0915aaa000)
        libQt5Network.so.5 => /usr/lib/libQt5Network.so.5 (0x00007f0915929000)
        libQt5Gui.so.5 => /usr/lib/libQt5Gui.so.5 (0x00007f091525c000)
        libQt5Core.so.5 => /usr/lib/libQt5Core.so.5 (0x00007f0914d0a000)
        libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x00007f0914b2b000)
        libm.so.6 => /usr/lib/libm.so.6 (0x00007f09149e5000)
        libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x00007f09149cb000)
        libc.so.6 => /usr/lib/libc.so.6 (0x00007f0914802000)
        libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00007f09147e0000)
        libpulse.so.0 => /usr/lib/libpulse.so.0 (0x00007f091478b000)
        libz.so.1 => /usr/lib/libz.so.1 (0x00007f091476f000)
        libdl.so.2 => /usr/lib/libdl.so.2 (0x00007f0914769000)
        libgssapi_krb5.so.2 => /usr/lib/libgssapi_krb5.so.2 (0x00007f0914714000)
        libssl.so.1.1 => /usr/lib/libssl.so.1.1 (0x00007f0914684000)
        libcrypto.so.1.1 => /usr/lib/libcrypto.so.1.1 (0x00007f09143a6000)
        libGL.so.1 => /usr/lib/libGL.so.1 (0x00007f0914320000)
        libpng16.so.16 => /usr/lib/libpng16.so.16 (0x00007f09142e7000)
        libharfbuzz.so.0 => /usr/lib/libharfbuzz.so.0 (0x00007f0914218000)
        libmd4c.so.0 => /usr/lib/libmd4c.so.0 (0x00007f0914202000)
        libsystemd.so.0 => /usr/lib/libsystemd.so.0 (0x00007f091414f000)
        libdouble-conversion.so.3 => /usr/lib/libdouble-conversion.so.3 (0x00007f0914138000)
        libicui18n.so.67 => /usr/lib/libicui18n.so.67 (0x00007f0913e38000)
        libicuuc.so.67 => /usr/lib/libicuuc.so.67 (0x00007f0913c4f000)
        libpcre2-16.so.0 => /usr/lib/libpcre2-16.so.0 (0x00007f0913bc4000)
        libzstd.so.1 => /usr/lib/libzstd.so.1 (0x00007f0913af2000)
        libglib-2.0.so.0 => /usr/lib/libglib-2.0.so.0 (0x00007f09139be000)
        /lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007f0915ed5000)
        libpulsecommon-13.99.so => /usr/lib/pulseaudio/libpulsecommon-13.99.so (0x00007f0913939000)
        libdbus-1.so.3 => /usr/lib/libdbus-1.so.3 (0x00007f09138e2000)
        libkrb5.so.3 => /usr/lib/libkrb5.so.3 (0x00007f09137fc000)
        libk5crypto.so.3 => /usr/lib/libk5crypto.so.3 (0x00007f09137cb000)
        libcom_err.so.2 => /usr/lib/libcom_err.so.2 (0x00007f09137c5000)
        libkrb5support.so.0 => /usr/lib/libkrb5support.so.0 (0x00007f09137b6000)
        libkeyutils.so.1 => /usr/lib/libkeyutils.so.1 (0x00007f09137af000)
        libresolv.so.2 => /usr/lib/libresolv.so.2 (0x00007f0913793000)
        libGLdispatch.so.0 => /usr/lib/libGLdispatch.so.0 (0x00007f09136dc000)
        libGLX.so.0 => /usr/lib/libGLX.so.0 (0x00007f09136a9000)
        libfreetype.so.6 => /usr/lib/libfreetype.so.6 (0x00007f09135d5000)
        libgraphite2.so.3 => /usr/lib/libgraphite2.so.3 (0x00007f09135b0000)
        librt.so.1 => /usr/lib/librt.so.1 (0x00007f09135a3000)
        liblzma.so.5 => /usr/lib/liblzma.so.5 (0x00007f091357b000)
        liblz4.so.1 => /usr/lib/liblz4.so.1 (0x00007f0913559000)
        libgcrypt.so.20 => /usr/lib/libgcrypt.so.20 (0x00007f0913438000)
        libicudata.so.67 => /usr/lib/libicudata.so.67 (0x00007f0911921000)
        libpcre.so.1 => /usr/lib/libpcre.so.1 (0x00007f09118af000)
        libsndfile.so.1 => /usr/lib/libsndfile.so.1 (0x00007f0911831000)
        libxcb.so.1 => /usr/lib/libxcb.so.1 (0x00007f0911807000)
        libasyncns.so.0 => /usr/lib/libasyncns.so.0 (0x00007f09117ff000)
        libX11.so.6 => /usr/lib/libX11.so.6 (0x00007f09116be000)
        libbz2.so.1.0 => /usr/lib/libbz2.so.1.0 (0x00007f09116ab000)
        libgpg-error.so.0 => /usr/lib/libgpg-error.so.0 (0x00007f0911682000)
        libFLAC.so.8 => /usr/lib/libFLAC.so.8 (0x00007f0911643000)
        libogg.so.0 => /usr/lib/libogg.so.0 (0x00007f0911638000)
        libvorbis.so.0 => /usr/lib/libvorbis.so.0 (0x00007f091160a000)
        libvorbisenc.so.2 => /usr/lib/libvorbisenc.so.2 (0x00007f091155f000)
        libXau.so.6 => /usr/lib/libXau.so.6 (0x00007f091155a000)
        libXdmcp.so.6 => /usr/lib/libXdmcp.so.6 (0x00007f0911550000)
```

**Comment by KitsuneRal** [comment]
Ok, so to sum up so far:
- Neochat is unstable, registering some event types twice; linked dynamically.
- Quotest runs fine, both when linked statically and dynamically; I assume it only registers each event type once.

Next step would be to build and run Spectral as the code closest to Neochat. Can you do it?

**Comment by derfenix** [comment]
Spectral version from aur didn't built for me because of non-related errors. But new version of the neochat not crashed any more (cbeafdd → 2a0f89b). So, I don't know if there is still reason to debug this thing. Guess it was neochat's fault, if it was fixed.

**Comment by KitsuneRal** [comment]
Alright. I didn't look closely into the changes, so no idea. Please reopen this if encountered again.

**Comment by TobiasFella** [comment]
Sorry, but i have to reopen this since a few more people reported the same problem. Only people building from AUR have this problem, so it's probably something with arch's compiler flags. I'll setup an arch system later to test myself

**Comment by KitsuneRal** [comment]
You haven't quite reopened it though; but ok, I'll do it now. I'm keeping the `invalid` label though, since instantiating the same template twice with the same arguments is against the C++ standard. If any combination of compiler flags allows it, it should be something very bad. Basically, you can't get any serious C++ project (including Qt itself) running if that point doesn't hold.

Independently, however, I'm working on the revised events infrastructure that will rule out even such invalid case. I might still leave a `Q_ASSERT` that will catch such occasions because it's abnormal no matter how you slice it; but at least Release builds will survive.

**Comment by TobiasFella** [comment]
(Sorry if I'm annoying you at this point :smile: )

We've gotten a few more reports of this happening on postmarketOS (=alpine linux) and fedora. In all cases, it crashes only with libQuotient from package managers and not when self-built.

I investigated a bit and found out that it's caused by [this](https://invent.kde.org/frameworks/extra-cmake-modules/-/blob/master/kde-modules/KDECompilerSettings.cmake#L220) line in extra-cmake-modules, setting CMP0063 to `new`. when i manually set it to `old` in my code, it doesn't crash anymore. The documentation says about this line

> No sane project should be affected by CMP0063, so suppress the warnings generated by the above visibility settings in CMake >= 3.3

Setting the policy to `old` is a good enough fix for us, so feel free to close this issue

**Comment by KitsuneRal** [comment]
That kinda puts libQuotient to the list of insane projects :-D thanks for investigating; it does look a bit more complicated than I thought originally. I'll keep it open for now.

**Comment by KitsuneRal** [comment]
Aah, so you explicitly set visibility of dynamic symbols... Yeah, libQuotient is not quite prepared for such things, so the workaround is proper. But thanks for finding what seems to be the root cause.

**Comment by nicolasfella** [comment]
We noticed that removing the -fvisibility-inlines-hidden in https://github.com/quotient-im/libQuotient/blob/dev/CMakeLists.txt#L43 makes the crash go away. I still don't fully understand what's going on, but given that there's quite some inline/template stuff in event.h it kinda checks out

**Comment by nicolasfella** [comment]
I have a feeling that we end up with duplicate symbols (one coming from event.h included by quotient and one coming from including event.h from neochat) in the app and therefore get duplicate event registrations. I stiff can't tell anything for sure though

**Comment by KitsuneRal** [comment]
Well, it helps the compiler to find existing (inline) template method instantiations, though I'm surprised that it works across shared object boundaries.

**Comment by KitsuneRal** [comment]
As I still fail to reproduce it - I wonder if removing `inline` either from:

1) `EventTypeRegistry::initializeTypeId()` (the templated one) or
2) `typeId()` (only the generic one; if you drop `inline` from the `typeId<void>()` specialisation you'll get linkage errors)

(both in `event.h`) helps with the crash.

I'm going to axe the whole `EventTypeRegistry` thing anyway; but it would be great to pinpoint which specific usage leads to the fallout.

**Comment by TobiasFella** [comment]
No, neither of those fixes the crash

**Comment by KitsuneRal** [comment]
Did you try both together?

**Comment by TobiasFella** [comment]
Yes

**Comment by KitsuneRal** [comment]
Alright, thanks... But at least do you not see repeated factory method additions for the same type as of the latest `dev`?

**Comment by KitsuneRal** [comment]
Summing up what we went through in the chat: the methods are still added two times to what seems to be duplicating factories. Comparison of exported symbols didn't find relevant differences between binaries on Tobias' machine and on mine. That exercise highlighted though that there are a lot of symbols defined as local that should not be there at all; and a lot of symbols defined as local that should actually be exported.

**Comment by KitsuneRal** [comment]
The last day's investigations brought up that it's just NeoChat double-registering but not Quotest or Quaternion. One possibly relevant difference is that NeoChat defines a custom `StickerEvent` while Quotest and Quaternion don't extend on the set of events provided by the library. [This branch](https://github.com/quotient-im/libQuotient/tree/kitsune/cleanup) adds a custom event definition and a test to Quotest - however, I still don't observe double registration with it (the test is valuable anyway). _Update_: and CI is happy too...

**Comment by KitsuneRal** [comment]
To report on the last day: long story short, `-fvisibility-inlines-hidden` does have something to do with all this - perhaps importantly, Quaternion and Quotest are compiled with this option unset while NeoChat (unless `CMP0063` is set to `OLD` - meaning that visibility presets only work on shared libraries) hides its inlines (including those defined in Quotient headers) as well. To summarise:
- if Quotient is compiled without `-fvisibility-inlines-hidden` or if `CMP0063` is set to `OLD` for NeoChat (I'm not quite sure why Quaternion/Quotest don't even warn about it, even when I `set(CMAKE_CXX_VISIBILITY_INLINES_HIDDEN ON)`): I venture guessing that it is enough for the dynamic loader to have inline symbols of the same name visible either in Quotient or NeoChat in order to deduplicate them
- If both Quotient and Neochat hide their inline symbols (and there seems to be something more that ECM does), the matching doesn't occur and each side runs its own initialisation code.

The (tentative) conclusion from all this is that I'm going to introduce proper visibility markup as I planned anyway; and then we'll check whether the problem still persists or that was the actual cause. Thanks everyone for patience and help so far.

**Comment by KitsuneRal** [comment]
News from these 24 hours - after applying visibility attributes throughout the library and flipping `CXX_VISIBILITY_PRESET` to `hidden` in CMake I can now reproduce the original issue with duplicate event typeIds in quotest (factories only get added once though - hopefully meaning that this part is done right). Further work to fix it is tracked in #529 from now.
